Assets/Thirdweb/Examples/Scripts/Prefabs/Prefab_Events.cs

```diff
@@ -0,0 +1,177 @@
+                "goerli",
+                "0x2e01763fA0e15e07294D74B63cE4b526B321E389"
+            );
+
+            // Optional event query options
+            EventQueryOptions options = new EventQueryOptions(null, 0, 16500000, "desc");
+
+            List<ContractEvent<object>> allContractEvents = await contract.events.GetAll(options);
+
+            foreach (ContractEvent<object> contractEvent in allContractEvents)
+                Debug.Log($"{contractEvent.ToString()}\n");
+        }
+        catch (System.Exception e)
+        {
+            Debug.Log($"Error: {e.Message}");
+        }
+    }
+
+    // Event listeners
+
+    public async void AddEventListener()
+    {
+        try
+        {
+            Contract contract = new Contract(
+                "goerli",
+                "0x2e01763fA0e15e07294D74B63cE4b526B321E389"
+            );
+
+            string result = await contract.events.AddListener(
+                "Transfer",
+                "Prefab_Events",
+                "OnTransfer"
+            );
+
+            Debug.Log($"Event listener added! Result: {result}");
+        }
+        catch (System.Exception e)
+        {
+            Debug.Log($"Error: {e.Message}");
+        }
+    }
+
+    public async void RemoveEventListener()
+    {
+        try
+        {
+            Contract contract = new Contract(
+                "goerli",
+                "0x2e01763fA0e15e07294D74B63cE4b526B321E389"
+            );
+
+            string result = await contract.events.RemoveListener(
+                "Transfer",
+                "Prefab_Events",
+                "OnTransfer"
+            );
+
+            Debug.Log($"Event listener removed! Result: {result}");
+        }
+        catch (System.Exception e)
+        {
+            Debug.Log($"Error: {e.Message}");
+        }
+    }
+
+    public async void ListenToAllEvents()
+    {
+        try
+        {
+            Contract contract = new Contract(
+                "goerli",
+                "0x2e01763fA0e15e07294D74B63cE4b526B321E389"
+            );
+
+            string result = await contract.events.ListenToAll("Prefab_Events", "OnAnyEvent");
+
+            Debug.Log($"Listening to all events! Result: {result}");
+        }
+        catch (System.Exception e)
+        {
+            Debug.Log($"Error: {e.Message}");
+        }
+    }
+
+    public async void RemoveAllEventListeners()
+    {
+        try
+        {
+            Contract contract = new Contract(
+                "goerli",
+                "0x2e01763fA0e15e07294D74B63cE4b526B321E389"
+            );
+
+            string result = await contract.events.RemoveAllListeners();
+
+            Debug.Log($"Removed all event listeners! Result: {result}");
+        }
+        catch (System.Exception e)
+        {
+            Debug.Log($"Error: {e.Message}");
+        }
+    }
+
+    public void OnTransfer()
+    {
+        Debug.Log("[EventListener] Transfer event was just emitted!");
```

Comment:
how would I get the emitted event here? don't think that works with the message passing approach since we need to pass data in the callback ?

Reply:
How about OnTransfer(ContractEvent event) and then on the JS side we always pass the event as the first argument in SendMessage? All callbacks that add a ContractEvent listener must have their first argument as the ContractEvent.

```js
contract.events.addEventListener("Transfer", (event) => {
  MyGameInstance.SendMessage(Prefab_Events, "OnTransfer", event, "OnTransferArg2", "OnTransferArg3", ...);
});
```